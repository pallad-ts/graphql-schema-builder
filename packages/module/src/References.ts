export const References = {
    SCHEMA: '@pallad/graphql-schema'
}