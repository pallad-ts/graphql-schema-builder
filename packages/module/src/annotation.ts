import {createAnnotationFactory} from "alpha-dic";

export const annotation = createAnnotationFactory('graphql-schema-builder', (order?: number) => ({order: order ?? 10}));

export type Annotation = ReturnType<typeof annotation>;