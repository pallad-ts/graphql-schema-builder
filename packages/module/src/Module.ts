import {Module as _Module, StandardActions} from '@pallad/modules';
import {Container} from "alpha-dic";
import {References} from "./References";
import {buildSchema, SchemaBuilder, SchemaComposer} from "@pallad/graphql-schema-builder";
import {annotation, Annotation} from "./annotation";
import * as is from 'predicates';
import {TypeRef} from "alpha-dic/compiled/TypeRef";
import {GraphqlSchemaBuilderShape} from "./GraphqlSchemaBuilderShape";


const isBuilderShape = is.struct({
    build: is.func
});

function hasBuilderShape(value: any): value is GraphqlSchemaBuilderShape {
    return isBuilderShape(value);
}

export class Module extends _Module<{ container: Container }> {
    constructor(private schemaComposer?: SchemaComposer<any>) {
        super('@pallad/graphql-schema-builder-module');
    }

    init() {
        this.registerAction(StandardActions.INITIALIZATION, ({container}) => {
            let typeRef = TypeRef.createFromType(SchemaComposer)!;
            if (this.schemaComposer) {
                const typeRefFromValue = TypeRef.createFromValue(this.schemaComposer);
                if (typeRefFromValue) {
                    typeRef = typeRefFromValue;
                }
            }

            container.definitionWithFactory(References.SCHEMA, async () => {
                const buildersWithAnnotations: Array<[SchemaBuilder | GraphqlSchemaBuilderShape, Annotation]> = await container.getByAnnotation(annotation.predicate, true);
                const builders = buildersWithAnnotations.sort(([, ann1], [, ann2]) => {
                    return ann1.order - ann2.order;
                })
                    .map(([builder]) => {
                        if (is.func(builder)) {
                            return builder;
                        }

                        if (hasBuilderShape(builder)) {
                            return builder.build.bind(builder);
                        }

                        throw new Error('One of Graphql schema builders is not a function or has a shape for GraphqlSchemaBuilderShape');
                    });
                return buildSchema(builders, this.schemaComposer);
            })
                .markType(typeRef);
        });
    }
}