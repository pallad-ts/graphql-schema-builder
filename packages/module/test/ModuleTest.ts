import {Engine, StandardActions} from "@pallad/modules";
import {Container, create} from "alpha-dic";
import {Module} from "@src/Module";
import * as sinon from 'sinon';
import {annotation} from "@src/annotation";
import {References} from "@src/References";
import {TypeRef} from "alpha-dic/compiled/TypeRef";
import {SchemaComposer} from "@pallad/graphql-schema-builder";

describe('Module', () => {

    let schemaBuilder1: sinon.SinonStub;
    let schemaBuilder2: {build: sinon.SinonStub};
    let schemaBuilder3: sinon.SinonStub;
    let container: Container;
    let engine: Engine<{container: Container}>;

    beforeEach(() => {
        container = create();
        engine = new Engine({container});
        schemaBuilder1 = sinon.stub();
        schemaBuilder2 = {
            build: sinon.stub()
        };

        schemaBuilder3 = sinon.stub();

        container.definitionWithValue(schemaBuilder1)
            .annotate(annotation(100));

        container.definitionWithValue(schemaBuilder2)
            .annotate(annotation());

        container.definitionWithValue(schemaBuilder3)
            .annotate(annotation(-10));
    })

    it('basic test', async () => {
        engine.registerModule(new Module());
        await engine.runAction(StandardActions.INITIALIZATION);
        const result = await container.get(References.SCHEMA);


        const schemaComposerMatch = sinon.match.instanceOf(SchemaComposer);
        sinon.assert.calledWithMatch(
            schemaBuilder1,
            schemaComposerMatch
        );

        sinon.assert.calledWithMatch(
            schemaBuilder2.build,
            schemaComposerMatch
        );

        sinon.assert.calledWithMatch(
            schemaBuilder3,
            schemaComposerMatch
        );

        sinon.assert.callOrder(
            schemaBuilder3,
            schemaBuilder2.build,
            schemaBuilder1
        );

        const def = container.findByName(References.SCHEMA);

        expect(def!.type)
            .toEqual(TypeRef.createFromType(SchemaComposer));

        expect(result)
            .toBeInstanceOf(SchemaComposer);
    });

    it('custom schema composer', async () => {
        const composer = new SchemaComposer();
        engine.registerModule(new Module(composer));
        await engine.runAction(StandardActions.INITIALIZATION);
        const result = await container.get(References.SCHEMA);

        sinon.assert.calledWithExactly(
            schemaBuilder1,
            composer
        );

        sinon.assert.calledWithExactly(
            schemaBuilder2.build,
            composer
        );

        sinon.assert.calledWithExactly(
            schemaBuilder3,
            composer
        );

        sinon.assert.callOrder(
            schemaBuilder3,
            schemaBuilder2.build,
            schemaBuilder1
        );

        expect(result)
            .toStrictEqual(composer);
    });

    it('throws error if once of schema builder is not a function or has a shape of builder', async () => {

        container.definitionWithValue({foo: 'bar'})
            .annotate(annotation());

        engine.registerModule(new Module());
        await engine.runAction(StandardActions.INITIALIZATION);

        return expect(container.get(References.SCHEMA))
            .rejects
            .toThrowErrorMatchingSnapshot();
    })
});