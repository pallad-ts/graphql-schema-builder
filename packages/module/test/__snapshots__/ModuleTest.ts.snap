// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module throws error if once of schema builder is not a function or has a shape of builder 1`] = `"One of Graphql schema builders is not a function or has a shape for GraphqlSchemaBuilderShape"`;
